onboard/catkin_ws/src/offboard_comms/scripts/arduino.py

```diff
@@ -0,0 +1,441 @@
+            if no_linebreaks:
+                print(port, end='')
+            else:
+                print(f'{OUTPUT_PREFIX}: {arduino_name.capitalize()}: {port}')
+        except StopIteration:
+            print(f'{OUTPUT_PREFIX}: Could not find port of {arduino_name} arduino.')
+
+
+def compile(arduino_names: List[str], print_output: bool) -> None:
+    """
+    Install required libraries and compile the sketches for the given Arduino names.
+
+    Args:
+        arduino_names: List of Arduino names.
+        print_output: Whether to allow the commands to print to stdout.
+    """
+
+    # Install the required libraries for the given Arduino names
+    install_libs(arduino_names, print_output)
+
+    # Print a linebreak
+    print()
+
+    # Compile the sketches for the given Arduino names
+    for index, arduino_name in enumerate(arduino_names):
+        print(f'{OUTPUT_PREFIX}: Compiling {arduino_name} arduino sketch...')
+
+        sketch_path_absolute = get_arduino_sketch_path_absolute(arduino_name)
+        fqbn = ARDUINO_DATA[arduino_name]['fqbn']
+
+        # Run pre-compile command, if it exists
+        if ARDUINO_DATA[arduino_name].get('pre_compile'):
+            run_command(ARDUINO_DATA[arduino_name]['pre_compile'], print_output)
+
+        # Compile the sketch
+        run_command(ARDUINO_COMPILE_COMMAND_TEMPLATE.format(fqbn=fqbn, sketch_path=sketch_path_absolute), print_output)
+
+        # Run post-compile command, if it exists
+        if ARDUINO_DATA[arduino_name].get('post_compile'):
+            run_command(ARDUINO_DATA[arduino_name]['post_compile'], print_output)
+
+        print(f'{OUTPUT_PREFIX}: Compilation of {arduino_name} arduino sketch complete.')
+
+        # Print a linebreak between arduino compilations
+        if index < len(arduino_names) - 1:
+            print()
+
+
+def upload(arduino_names: List[str], print_output: bool) -> None:
+    """
+    Install required libraries, compile the sketches, and upload them to the given Arduino names.
+
+    Args:
+        arduino_names: List of Arduino names.
+        print_output: Whether to allow the commands to print to stdout.
+    """
+
+    # Compile the sketches for the given Arduino names
+    compile(arduino_names, print_output)
+
+    # Print a linebreak
+    print()
+
+    print(f'{OUTPUT_PREFIX}: Finding ports of the given arduinos...')
+
+    # Get the list of ports for the given Arduino names
+    ports = {}
+    error = False
+    for arduino_name in arduino_names:
+        try:
+            ports[arduino_name] = get_arduino_port(arduino_name)
+            print(f'{OUTPUT_PREFIX}: {arduino_name.capitalize()}: {ports[arduino_name]}')
+        except StopIteration:
+            error = True
+            print(f'{OUTPUT_PREFIX}: FATAL ERROR: Could not find port of {arduino_name} arduino.')
+
+    if error:
+        print(f'{OUTPUT_PREFIX}: Could not upload sketches to arduinos due to errors in finding ports '
```

Comment:
Would it be more appropriate to raise an exception here?

Reply:
Yes, especially if this function would be called from another function. However, that is not the case here, and is very unlikely to be the case in the future. The downside with raising an exception is that it prints the traceback, which is not helpful here and would be confusing for anyone not programming-oriented. There also doesn't seem to be a good way to raise an exception in Python without a traceback. Thus, printing the error message and simply returning keeps the output clean and coveys the message without being too confusing.